src/openassetio-python/bridge/src/python/converter.cpp

```diff
@@ -0,0 +1,105 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2022 The Foundry Visionmongers Ltd
+#include <openassetio/python/converter.hpp>
+
+#include <pybind11/embed.h>
+
+#include <openassetio/Context.hpp>
+#include <openassetio/TraitsData.hpp>
+#include <openassetio/hostApi/HostInterface.hpp>
+#include <openassetio/hostApi/Manager.hpp>
+#include <openassetio/hostApi/ManagerFactory.hpp>
+#include <openassetio/hostApi/ManagerImplementationFactoryInterface.hpp>
+#include <openassetio/log/ConsoleLogger.hpp>
+#include <openassetio/log/LoggerInterface.hpp>
+#include <openassetio/log/SeverityFilter.hpp>
+#include <openassetio/managerApi/Host.hpp>
+#include <openassetio/managerApi/HostSession.hpp>
+#include <openassetio/managerApi/ManagerInterface.hpp>
+#include <openassetio/managerApi/ManagerStateBase.hpp>
+
+// Private headers
+#include <openassetio/private/python/pointers.hpp>
+
+namespace py = pybind11;
+
+namespace openassetio {
+inline namespace OPENASSETIO_CORE_ABI_VERSION {
+namespace python::converter {
+
+template <typename T>
+PyObject* castToPyObject(const T& objectPtr) {
+  // Stash the errors for restoring after cast.
+  // An error_scope calls PyErr_Fetch (which clears the active error),
+  // and stashes them, restoring the error state when it falls out of
+  // scope
+  const py::error_scope previousErrorState;
+
+  if (objectPtr == nullptr) {
+    throw std::invalid_argument("objectPtr cannot be null");
+  }
+
+  py::object pybindObj = py::cast(objectPtr);
+
+  // Check to see if the py::cast has spawned a python error
+  // We use another error_scope here for convenience of access, relying
+  // on the fact that the original error scope will be destructed last,
+  // resetting to the initial error.
+  const py::error_scope castErrorState;
+
+  if (castErrorState.value != nullptr) {
+    throw std::runtime_error(py::str(castErrorState.value));
+  }
+
+  // `release()` to avoid decrementing the PyObject refcount on leaving
+  // this function. We explicitly want +1 refcount.
+  return pybindObj.release().ptr();
+}
+
+template <typename T>
+typename T::Ptr castFromPyObject(PyObject* pyObject) {
+  if (pyObject == nullptr) {
+    throw std::invalid_argument("Attempting to cast a nullptr PyObject");
+  }
+
+  // get Python object.
```

Comment:
[fix] (unless its obvious/the only way) can we explain why `reinterpret_borrow` specifically?

Reply:
I'm not actually sure, I took this from the other method that uses `PyRetainingSharedPtr`.

From my breif pybind foray, I think we could replace this with `std::cast<py::object>(pyObject);` and it would do the same thing, all the tests pass if you do.

@feltech What's your thoughts?

Reply:
One option that could sidestep this issue is to replace with
```
const auto pyInstance = py::handle(pyObject);
```

Reply:
All that said, is there anything to say here? We're not really making a choice to borrow, we're just making a choice not to steal, which is the default choice. If this was written as `py::cast<py::object>` would we be having this discussion?

Reply:
fine to leave it if its the obvious choice

Reply:
Happy with whatever you folks decide, closing this from a docs perspective.